bson/bsoncodec/bsoncodec.go

```diff
@@ -119,13 +119,63 @@ func (vde ValueDecoderError) Error() string {
 // value.
 type EncodeContext struct {
```

Comment:
It would probably be too complex for the scope of GODRIVER-2716, but what are your thoughts on creating a 2.0 ticket to un-export the EncodeContext? This wouldn't limit a user from defining a Readers/Writers for defining a custom decoder, but it would prevent users from creating custom value encoders. 

The would make the codec less vulnerable for misuse. 

If not the EncodeContext, what are your thoughts on un-exporting the individual fields?

Same question for DecodeContext.

Reply:
I think we still need to support defining custom codecs using a Registry in Go Driver 2.0. We could remove `EncodeContext`/`DecodeContext` from the `ValueEncoder`/`ValueDecoder` function signatures, which would simplify the API, but would remove the ability to access a `Registry` from user-defined encoders/decoders. It's not clear what effect that would have on implementing user-defined encoders/decoders (e.g. would it make some impossible to implement?).

I think un-exporting the new `EncodeContext` and `DecodeContext` boolean fields is a good idea. The downside is that we still need to use `EncodeContext`/`DecodeContext` to communicate to the built-in codecs when a user has set configurations on a `bson.Encoder` or `bson.Decoder`, so we still need configuration methods on `EncodeContext` and `DecodeContext` that can be called from the `bson` package. That approach follows the existing pattern used by the `DefaultDocumentM` and `DefaultDocumentD` methods. I've updated the PR to un-export the new boolean configuration values in `EncodeContext` and `DecodeContext` and add exported methods for setting them.

Reply:
In Go Driver 2.0, the only use of `EncodeContext`/`DecodeContext` will be to provide access to the encoder/decoder registry from a custom `ValueEncoder` or `ValueDecoder`.